### PR TITLE
Add support for ESP internal temperature sensor

### LINKED ISSUE #333: Missing support for internal temperature sensor
Various ESP32 models have an integrated temperature sensor that is supported by [esp-idf](https://github.com/espressif/esp-idf). 
It would be nice to see support for that sensor built into this crate.

I would like to try implement it - what steps need to be taken for such an implementation? Is there documentation for that process? Is this crate even the right place or does it need to be implemented in [esp-idf-sys](https://github.com/esp-rs/esp-idf-sys) instead?

#### Links

- [Documentation (for sensor on ESP32C3)](https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/api-reference/peripherals/temp_sensor.html)
- [Example usage in esp-idf](https://github.com/espressif/esp-idf/blob/master/examples/peripherals/temperature_sensor/temp_sensor/main/temp_sensor_main.c)
- [Driver in esp-dif](https://github.com/espressif/esp-idf/blob/master/components/driver/temperature_sensor/temperature_sensor.c)

### PR DESCRIPTION
**Pull request by GamerBene19**
What the title says. Currently very much WIP. PR created to help give feedback.

Closes #333

### CODE REVIEW

**Comment by GamerBene19** [comment]
@Vollbrecht FYI

File: src/temp_sensor.rs
Comment (Vollbrecht) [inline src/temp_sensor.rs:2]: when you import a crate its not automatically scoped to all the child modules inside the file

File: src/temp_sensor.rs
Comment (Vollbrecht) [inline src/temp_sensor.rs:12]: you need to import here with either `use super::*` or the better way the module or directly move the above use inside the module so you dont need extra scopes

Comment (GamerBene19) [reply src/temp_sensor.rs:12]: I see. That was the missing information. Thanks!

File: src/temp_sensor.rs
Comment (GamerBene19) [reply src/temp_sensor.rs:2]: https://github.com/esp-rs/esp-idf-hal/pull/337#discussion_r1391807390

File: src/temp_sensor.rs
Comment (Vollbrecht) [inline src/temp_sensor.rs:14]: to make the #[cfg(any(esp32s2, esp32s3, esp32c2, esp32c3, esp32c6, esp32h2))] not so verbose i think its currently best to list the "not any" - so the original esp32 and so forth, because this list will potentially only grow and the chips that not support it will be more likely the same

File: src/temp_sensor.rs
Comment (Vollbrecht) [inline src/temp_sensor.rs:104]: to have an consistent naming schema across the modules, the main struct that impl all the methonds and do the stuff has a Driver in the name - PinDriver, SpiDriver , TemperatureSensorDriver, its more verbose but consistend

File: src/temp_sensor.rs
Comment (Vollbrecht) [inline src/temp_sensor.rs:113]: std::ptr::null_mut has a counterpart -> core::ptr::null_mut, we do that for the reasons i stated in the issue

File: Cargo.toml
Comment (Vollbrecht) [inline Cargo.toml:37]: you can now directly reference your branch of this on github with git = and branch= , this will give the ci the chance to run correctly

**Review by Vollbrecht — CHANGES_REQUESTED** [review]
i added some comments, piece by piece we will get there.

**Comment by Vollbrecht** [comment]
The interesting point is when you start implement the callback [functions](https://github.com/espressif/esp-idf/blob/c8243465e45/components/driver/temperature_sensor/include/driver/temperature_sensor.h#L180)  provided by the driver. It should allow you to setup a callback on certain temperature condition. In our crate terminology we call it a subscribe, if you want to peak in other drivers how similar things are handled.

File: src/temp_sensor.rs
Comment (GamerBene19) [reply src/temp_sensor.rs]: Done in bb519ff751c16191eebb1f54592bd8ed43b72d68

File: src/temp_sensor.rs
Comment (GamerBene19) [reply src/temp_sensor.rs]: Done in 04bd17366920c84be6086891c4aeb99e4ceca24a

File: src/temp_sensor.rs
Comment (GamerBene19) [reply src/temp_sensor.rs]: Done in e9eff88211fd42d2c2d896110d55eec0cb32dbd4

File: Cargo.toml
Comment (GamerBene19) [reply Cargo.toml]: Done in 6f9a02fc6fdf545ebec16e27d07532b97f8753dd

File: src/temp_sensor.rs
Comment (Vollbrecht) [inline src/temp_sensor.rs:115]: don,'t eat the errors you get back from the underlying C api calls, every Error should be propagated. You can use a `Result<(),EspError> on the return's of all the method calls where its needed

Comment (Vollbrecht) [reply src/temp_sensor.rs:115]: for error converting we have the esp! macro

File: src/temp_sensor.rs
Comment (Vollbrecht) [inline src/temp_sensor.rs:161]: self.disable is not an unsafe call - try to scope unsafe to the minimal part where it is originated from

File: src/temp_sensor.rs
Comment (GamerBene19) [reply src/temp_sensor.rs]: Done in 2b0694d067e75c1ffbe118c3a890e6b6801fb5c8. The only thing I'm still uncertain about is how to handle errors that occur in drop, since drop has to return (). Perhaps you have a good idea.

File: src/temp_sensor.rs
Comment (GamerBene19) [reply src/temp_sensor.rs]: Done in 2b0694d067e75c1ffbe118c3a890e6b6801fb5c8

File: src/temp_sensor.rs
Comment (Vollbrecht) [reply src/temp_sensor.rs]: yeah you don't have many chances in a drop. We either eat the error, unwrap it or panic if we get one. The underlying driver probably still sends something out over Error println. Out of my head in most cases we eat the error in the othere drivers here.

**Comment by Vollbrecht** [comment]
If you are unsure how to proceed now to get the rest of the exposed api integrated, we can wrap it up here and i can integrate the rest ontop of it, or if you feel advantageous we can go deeper getting the callbacks to work with the api. 
I am not sure whats your background  so i leave it up to you.

**Comment by GamerBene19** [comment]
> since it seams to work with the flags we are now using in esp-idf-sys. We could also change it here. Rustcompiler give us the flag as `esp_idf_soc_temp_sensor_supported`.

Would that then be `#[cfg(esp_idf_soc_temp_sensor_supported)]`?

**Comment by jwhitlark** [comment]
Is there any further progress on this? I’m having problems with a chip overheating and would like to quantify the problem better.

**Comment by lu-zero** [comment]
Since I cannot update the PR myself I opened a rebase of it.

**Comment by GamerBene19** [comment]
> Is there any further progress on this? I’m having problems with a chip overheating and would like to quantify the problem better. Usable basic support would be helpful, if other work is paused. I’d be happy to help out, although I’m new enough to this stuff I don’t know how much use I’d be.

Sorry for the late reply, I forgot to answer back then.

As from my side: I might find some time to work on this in the next few weeks, but I likely need some time to get back into it (which repo of esp-rs does what, what is still missing, ...) and some more help (see question above). If someone else wants to do it instead, feel free :)

I'm pinging @ivmarkov too since he was asking about the state of the PR that @lu-zero opened.

**Comment by lu-zero** [comment]
I spent a bit of time to try to get it to at least pass the CI, I guess the remaining bits is to wire in the Peripheral and figure out if there are remaining failures.

**Comment by ivmarkov** [comment]
> I spent a bit of time to try to get it to at least pass the CI, I guess the remaining bits is to wire in the Peripheral and figure out if there are remaining failures.

I'm pretty sure we can finish the driver still today or so. It is so simple, that it needs an extra couple of hours at most. And I'll assist of course. :) What I can't help with is putting my own time to test it, but I'm hopeful you can do that.

**Comment by GamerBene19** [comment]
> I'm pretty sure we can finish the driver still today or so.

I'm back at home and have a few hours free time this evening. I'll try to address (some of) the remaining issues.

@lu-zero Do you mind if I cherry-pick your commits into my branch/this PR?

**Comment by lu-zero** [comment]
Please reset your tree to mine

**Comment by GamerBene19** [comment]
> Please reset your tree to mine, it should be the quickest way to take all I did and rebase at the same time

Done. Do you want to continue here then? If yes I'll give you write access to my repo so you can commit here too.

**Comment by lu-zero** [comment]
Sure, ideally what's missing is the Peripheral boilerplate that I wrote but I have yet to commit and wiring it in, and addressing @ivmarkov pending review items.

**Comment by lu-zero** [comment]
Looks like either -sys is not building the feature or the s2/s3 series do not support it.

**Comment by Vollbrecht** [comment]
You need to carefully check **for every esp variant** what underlying clock system is used, what options are there and what the default options, that is driving the underlying driver. E.g on some devices you can use the RC clock and the XTAL on other systems you may only use the RC clock and so forth.

You can simply check each definition by greping the esp-idf source and look for the coresponding `soc_periph_temperature_sensor_clk_src_t` definitions.

**Comment by GamerBene19** [comment]
> Sure, ideally what's missing is the Peripheral boilerplate that I wrote but I have yet to commit and wiring it in, and addressing @ivmarkov pending review items.

You should have access I think. Let me know if it does not work, than I can take another look.

**Comment by GamerBene19** [comment]
> You can simply check each definition by greping the esp-idf source and look for the coresponding soc_periph_temperature_sensor_clk_src_t definitions.

- C2, C3, C5, C6, C61, H2 variants all have support for both XTAL and RC_FAST, and use the former as their default
- P4 variant (which I'm not sure this library supports) does only have support for LP_PERI (and uses that as default)
- S2 and S3 only have support for RC_FAST, which also is their default.

I will try to commit it shortly.

Just a quick question: How can I build the esp-idf-hal project? A simple `cargo build` does fail. Iirc there was a way, but I forgot it.

**Comment by Vollbrecht** [comment]
> Just a quick question: How can I build the esp-idf-hal project? A simple `cargo build` does fail. Iirc there was a way, but I forgot it.

`ESP_IDF_VERSION={esp-idf-tag} MCU={mcu-name} cargo b --target {target-name} -Zbuild-std=std,panic_abort` // and optional features  directly provided

so for example for a c3 it lookse something like this: 
`ESP_IDF_VERSION=v5.2.2 MCU=esp32c3 cargo b --target riscv32imc-esp-espidf -Zbuild-std=std,panic_abort`

same can be done with `cargo clippy` etc

**Comment by GamerBene19** [comment]
> `ESP_IDF_VERSION=v5.2.2 MCU=esp32c3 cargo b --target riscv32imc-esp-espidf -Zbuild-std=std,panic_abort`

Thanks! I've switched to the nightly toolchain (for -Z) with `rustup override set nightly` and that your command which worked fine.
Unfortunately it fails for the S2 variant. I've tried 
`ESP_IDF_VERSION=v5.2.2 MCU=esp32s2 cargo clippy --target xtensa-esp32s2-espidf -Zbuild-std=std,panic_abort`
and got multiple `'esp32-s2' is not a recognized processor for this target (ignoring processor)` messages and what I assume to be the issue
```
error: data-layout for target `xtensa-esp32s2-espidf`, `e-m:e-p:32:32-v1:8:8-i64:64-i128:128-n32`, differs from LLVM target's `xtensa-none-elf` default layout, `e-m:e-p:32:32-i8:8:32-i16:16:32-i64:64-n32`
```

Do I need anything else for S2/S3 variants?

File: src/temp_sensor.rs
Comment (GamerBene19) [inline src/temp_sensor.rs:141]: @ivmarkov is this what you had in mind with

> Take the config by & as we do everywhere else.

?

Comment (ivmarkov) [reply src/temp_sensor.rs]: Yes. To avoid the `clone()` though you need to change `From<TemperatureSensorConfig>` to `From<&TemperatureSensorConfig>`.

Also, I'm no longer sure who is driving it - is that you, or @lu-zero ? Ultimately, we of course want one single PR that we can merge, so the two of you should probably sync with each other etc.. :)

Comment (GamerBene19) [reply src/temp_sensor.rs]: > Yes. To avoid the clone() though you need to change From<TemperatureSensorConfig> to From<&TemperatureSensorConfig>.

Alright, will do that (later).

> Also, I'm no longer sure who is driving it - is that you, or @lu-zero ? 

Both of us I think.

> Ultimately, we of course want one single PR that we can merge, so the two of you should probably sync with each other etc.. :)

As far as I understood https://github.com/esp-rs/esp-idf-hal/pull/337#issuecomment-2336828281, @lu-zero is fine  with continuing here. I've reset my branch to his and have also given him write access to my repo so he can add to this PR.

(@lu-zero Let me know if I misunderstood something)

Comment (GamerBene19) [reply src/temp_sensor.rs]: > Yes. To avoid the clone() though you need to change From<TemperatureSensorConfig> to From<&TemperatureSensorConfig>.

I've encountered a problem with `value.clk_src` in the From trait impl. As far as I understood it, `into()` consumes the value instead of taking a reference. The solution I ended up with (https://github.com/esp-rs/esp-idf-hal/pull/337/commits/c061d00195bdb167e796cae140d763192774c84d) is to change the trait for `TempSensorClockSource` too. I'm relatively new to Rust, so let me know if there is another/better way.

Comment (ivmarkov) [reply src/temp_sensor.rs]: > > Yes. To avoid the clone() though you need to change From to From<&TemperatureSensorConfig>.
> 
> I've encountered a problem with `value.clk_src` in the From trait impl. As far as I understood it, `into()` consumes the value instead of taking a reference. The solution I ended up with ([c061d00](https://github.com/esp-rs/esp-idf-hal/commit/c061d00195bdb167e796cae140d763192774c84d)) is to change the trait for `TempSensorClockSource` too. I'm relatively new to Rust, so let me know if there is another/better way.

Yes - just mark `TempSensorClockSource` as `Copy`. It is a simple "C style" enum so marking these with `Copy` is always a good idea.

File: src/temp_sensor.rs
Comment (ivmarkov) [inline src/temp_sensor.rs:86]: Add `#[non_exhaustive]` so that we have backwards compat with future extensions to this struct.

File: src/temp_sensor.rs
Comment (ivmarkov) [reply src/temp_sensor.rs]: No, we panic

File: src/temp_sensor.rs
Comment (GamerBene19) [reply src/temp_sensor.rs:86]: Done in https://github.com/esp-rs/esp-idf-hal/pull/337/commits/e32f03db07e5fe4cd63c093e3349011c4bae1478

**Comment by GamerBene19** [comment]
> > since it seams to work with the flags we are now using in esp-idf-sys. We could also change it here. Rustcompiler give us the flag as `esp_idf_soc_temp_sensor_supported`.
> 
> Would that then be `#[cfg(esp_idf_soc_temp_sensor_supported)]`?

Can (and should) we use this flag instead of `#[cfg(all(not(esp32)))]`?

File: src/temp_sensor.rs
Comment (GamerBene19) [reply src/temp_sensor.rs:1]: Resolved by https://github.com/esp-rs/esp-idf-hal/pull/337/commits/01ce494260da154db5794c89a94ec77bb9bdef14.

File: src/temp_sensor.rs
Comment (GamerBene19) [reply src/temp_sensor.rs]: Resolved by https://github.com/esp-rs/esp-idf-hal/pull/337/commits/01ce494260da154db5794c89a94ec77bb9bdef14.

**Comment by ivmarkov** [comment]
> @ivmarkov Regarding the peripheral, I've
> 
> * added `crate::impl_peripheral!(TempSensor);` in temp_sensor.rs
> * added
>   ```
>   #[cfg(all(not(esp32), esp_idf_version_major = "5"))]
>   pub temp_sensor: temp_sensor::TempSensor,
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   to Peripherals struct
>   and
>   ```
>   #[cfg(all(not(esp32), esp_idf_version_major = "5"))]
>   temp_sensor: temp_sensor::TempSensor::new(),
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   to it's impl
> * Adjusted the function signature of the driver constructor to be
>   ```
>   impl<'d> TempSensorDriver {
>   pub fn new(
>       config: &TempSensorConfig,
>       temp_periph: impl Peripheral<P = TempSensor> + 'd,
>       _p: PhantomData<&'d mut ()>,
>   ) -> Result<Self, EspError>
>   ```
> 
> What do I do with `temp_periph` now? Do I even need to use it, or can I simply prefix it with _?

It should be instead:
>   ```rust
>   impl<'d> TempSensorDriver<'d> {
>   pub fn new(
>       config: &TempSensorConfig,
>       _sensor: impl Peripheral<P = TempSensor> + 'd,
>   ) -> Result<Self, EspError>
>   ```
Notice:
- That the driver is now lifetimed by `'d`. And it should be
  ```rust
   pub struct TempSensorDriver<'d> {
       ptr: ...
       _p: PhantomData<&'d mut ()>,
   }
   ``` 
- The removal of `_p` from the constructor function. The `_p` is just a zero-sized trick so that the driver "remembers" that it might get the peripheral by `&'d mut` (i.e. the peripheral - when passed by `&mut` stays mutably borrowed until the driver is dropped)

Also, you need to initialize the `_p` in `TempSensorDriver` simply with ... `PhantomData`. No need to pass the `PhantomData` instance on constructor, as already mentioned above. That's an internal driver detail/trick.

**Comment by ivmarkov** [comment]
@GamerBene19 @lu-zero Once you feel this is ready (should be, once you address the non-panicing behavior of the dtor), please un-draft it so that we can see if it builds OK.

Post-merge I might implement the usual `subscribe` / `subscribe_nonstatic` dance as well as its more ergonomic `wait` `async` variant, so that folks don't have to constantly poll the sensor in a loo so as to detect when the device starts to over-heat...

File: src/temp_sensor.rs
Comment (GamerBene19) [reply src/temp_sensor.rs]: Done in https://github.com/esp-rs/esp-idf-hal/pull/337/commits/38a5156fac1dbb3e510b4947071b298b66411910

**Comment by GamerBene19** [comment]
> > > since it seams to work with the flags we are now using in esp-idf-sys. We could also change it here. Rustcompiler give us the flag as `esp_idf_soc_temp_sensor_supported`.
> > 
> > 
> > Would that then be `#[cfg(esp_idf_soc_temp_sensor_supported)]`?
> 
> Can (and should) we use this flag instead of `#[cfg(all(not(esp32)))]`?

@ivmarkov Do you have an opinion here?

**Comment by ivmarkov** [comment]
@GamerBene19 Oh one more thing:
```rust
unsafe impl<'d> Send for TempSensorDriver<'d> {}
```

This is necessary so that you can create the driver in one thread and then send it to another (as long as the periph is not mutably borrowed, but owned by the driver). `Send`ing the driver should be completely safe (even if the driver is not thread-safe) so why not?

**Comment by ivmarkov** [comment]
> > > > since it seams to work with the flags we are now using in esp-idf-sys. We could also change it here. Rustcompiler give us the flag as `esp_idf_soc_temp_sensor_supported`.
> > > 
> > > 
> > > Would that then be `#[cfg(esp_idf_soc_temp_sensor_supported)]`?
> > 
> > 
> > Can (and should) we use this flag instead of `#[cfg(all(not(esp32)))]`?
> 
> @ivmarkov Do you have an opinion here?

First of alla small nit: `#[cfg(all(not(esp32)))]` should be just `#[cfg(not(esp32))]` or else Clippy might kick-in.
With that said, either the above ^^^ or `#[cfg(esp_idf_soc_temp_sensor_supported)]` would work, up to you. Though I'm surprised that the `soc` flags are exposed. We are only supposed to support `CONFIG_` ESP IDF entries (and only boolean ones) as Rust `#[cfg]`s; not sure how the `_SOC_` stuff sneaked in, but yes, why not given that it is in there?

**Comment by ivmarkov** [comment]
@GamerBene19 @lu-zero I guess from your side this is ready for merge?

**Comment by lu-zero** [comment]
I'm happy with what is done and since I have access to this branch I can close my PR in favor of this one.

**Comment by GamerBene19** [comment]
> @GamerBene19 @lu-zero I guess from your side this is ready for merge?

Unfortunately I currently don't have a functioning ESP with which I could test this code. I think it _should_ work given that there were no "real" changes to the logic, but I think it makes sense to test it before merging.

Also I'd like to try using #[cfg(esp_idf_soc_temp_sensor_supported)] as it is more future proof, I've just pushed a commit to try it out.

**Comment by ivmarkov** [comment]
> > @GamerBene19 @lu-zero I guess from your side this is ready for merge?
> 
> Unfortunately I currently don't have a functioning ESP with which I could test this code. I think it _should_ work given that there were no "real" changes to the logic, but I think it makes sense to test it before merging.
> 
> Also I'd like to try using #[cfg(esp_idf_soc_temp_sensor_supported)] as it is more future proof, I've just pushed a commit to try it out.

@lu-zero Can you check the PR on an MCU?

**Comment by lu-zero** [comment]
Checked and seems working at least on the c3 and c6, added an example while at it. Shall we keep or remove the get_kelvin and get_fahrenheit?

**Comment by lu-zero** [comment]
Not sure why I cannot really push on the branch but here is the example used https://github.com/GamerBene19/esp-idf-hal/pull/1

**Comment by ivmarkov** [comment]
> Checked and seems working at least on the c3 and c6, added an example while at it. Shall we keep or remove the get_kelvin and get_fahrenheit?

Ah, keep them - not so important.

Can't see the example in the PR though?

**Comment by GamerBene19** [comment]
> Not sure why I cannot really push on the branch

As far as I can tell you should be able to, you have collaborator access and I don't see an option to explicitly grant you write permissions. 

> Can't see the example in the PR though?

It's here now

**Comment by ivmarkov** [comment]
> > Not sure why I cannot really push on the branch
> 
> As far as I can tell you should be able to, you have collaborator access and I don't see an option to explicitly grant you write permissions.
> 
> > Can't see the example in the PR though?
> 
> It's here now

Ah, CI failed. Look at the other examples as to how to build the example only for certain mcus.

**Comment by lu-zero** [comment]
Github does not let me push to that branch... @GamerBene19 pick the amended commit from https://github.com/lu-zero/esp-idf-hal/commit/9ce159f93dfef0956862327da8c99f8ffca88550

**Comment by lu-zero** [comment]
Sorry, pick the patch again, forgot to move the use inside the main.

**Comment by ivmarkov** [comment]
Thank you both! Merging.

**Comment by thomas725** [comment]
As far as I understood it the original ESP32 also has such an internal temperature sensor, but it's measurements accuracy is so bad that they decided to remove it from their datasheets. Is it possible to enable the API for the original ESP32 anyway or is it also missing from IDF and therefore enabling it in the Rust wrapper wouldn't work?

**Comment by Vollbrecht** [comment]
Yes its not possible via the normal temp api, since we relight on esp_idf_soc_temp_sensor_supported flag and that is set in esp-idf itself. Also even if there would be a way, we would not include it because its officially broken. So its not useful to include in a hal itself.

**Comment by thomas725** [comment]
thanks for your reply @Vollbrecht ! I found a surprisingly easy way to read the value by looking at how Tasmota does it:
```
extern "C" {
    fn temprature_sens_read() -> u_int8_t;
}
...
let msg = format!("CPU Temperature: {:?}", unsafe {temprature_sens_read()});
```

**Pull request merged**
